drivers/mcp9808/mcp9808.c

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (C) 2018 Inria
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_mcp9808
+ * @{
+ *
+ * @file
+ * @brief       Device driver implementation for the MCP9808 temperature sensor.
+ *
+ * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
+ *
+ * @}
+ */
+
+#include <math.h>
+
+#include "log.h"
+#include "mcp9808.h"
+#include "mcp9808_constants.h"
+#include "mcp9808_params.h"
+#include "periph/i2c.h"
+
+#define ENABLE_DEBUG        (0)
+#include "debug.h"
+
+#define DEV_I2C      (dev->params.i2c_dev)
+#define DEV_ADDR     (dev->params.i2c_addr)
+
+int _read16(const mcp9808_t *dev, uint8_t pointer, uint16_t *reg)
+{
+    /* Acquire exclusive access */
+    i2c_acquire(DEV_I2C);
+    uint8_t buf[2];
+    if (i2c_read_regs(DEV_I2C, DEV_ADDR, pointer, &buf, 2, 0) != 0) {
+        DEBUG("[mcp9808] Error: cannot read register 0x%02X\n", pointer);
+        i2c_release(DEV_I2C);
+        return -MCP9808_ERR_I2C;
+    }
+
+    *reg = (buf[0] << 8) | buf[1];
+
+    /* Release I2C device */
+    i2c_release(DEV_I2C);
+
+    return MCP9808_OK;
+}
+
+int _write16(const mcp9808_t *dev, uint8_t pointer, uint16_t *reg)
+{
+    uint8_t buf[2];
+    buf[0] = (uint8_t)((*reg) >> 8);
+    buf[1] = (uint8_t)(*reg);
+    if (i2c_write_regs(DEV_I2C, DEV_ADDR, pointer, &buf, 2, 0) != 0) {
+        DEBUG("[mcp9808] Error: cannot write register 0x%02X\n", pointer);
+        i2c_release(DEV_I2C);
+        return -MCP9808_ERR_I2C;
+    }
+
+    /* Release I2C device */
+    i2c_release(DEV_I2C);
+
+    return MCP9808_OK;
+}
+
+int mcp9808_init(mcp9808_t *dev, const mcp9808_params_t *params)
+{
+    dev->params = *params;
+
+    /* Acquire exclusive access */
+    i2c_acquire(DEV_I2C);
+
+    /* Check device ID */
+    uint8_t checkid;
+    i2c_read_reg(DEV_I2C, DEV_ADDR, MCP9808_REG_DEVICE_ID_REV, &checkid, 0);
+    if (checkid != MCP9808_DEVICE_ID) {
+        DEBUG("[mcp9808] Error: wrong device ID 0x%02X\n", checkid);
+        i2c_release(DEV_I2C);
+        return -MCP9808_ERR_NODEV;
+    }
+
+    /* Reset config register to default values */
+    uint16_t init = 0;
+    if (_write16(dev, MCP9808_REG_CONFIG, &init) != MCP9808_OK) {
```

Comment:
Here `_write16` is releasing the I2C bus and then it's being used again without acquiring. I think maybe the release in `_write16` could be removed and handled outside.